Gradient-Based Optimization from cas_simplification branch

nightdr opened the pull request:
Includes changes involving the implementation of gradient-based optimization methods for continuous local optimization. I started the architecture for this optimization and implemented it in the test cases and with explicit regression.

gbomarito commented on bingo/symbolic_regression/atomic_potential_regression.py:
```diff
@@ -52,6 +52,9 @@ def evaluate_fitness_vector(self, individual):
 
         return np.array(err_vec).flatten()
 
+    def get_jacobian(self, individual):
```
rather than having every fitness function be required to implement this function lets implement this with mixins:

1) make a GradientMixin class which has the abstract function for get_gradient

2) make a subclass of GradientMixin  called VectorGradientMixin.  this class has the get_gradient implementation that you currently have in the vectorbasedfitnessfunction  (including the abstract get_jacobian).

3) when getting ready to call scipy in the continuous local optimization you can test whether the fitnessfunction is a GradientMixin/VectorGradientMixin  and only use the gradient based omprimization in that case.

this also helps because then we don't have to modify all the implemented fitness functions, they just dont use the gradientMixin

nightdr replied:
sounds great, I did this in the new changes!

gbomarito reviewed (changes requested):
Pretty solid start, thanks for cleaning up some of the older stuff :)

nightdr commented on bingo/evaluation/fitness_function.py:
```diff
@@ -95,18 +98,57 @@ def __call__(self, individual):
+        fitness_vector = self.evaluate_fitness_vector(individual)
+        fitness_derivatives = self.get_jacobian(individual).transpose()
+
+        gradient = np.zeros(len(fitness_derivatives))
```
sounds good, I changed this method and the metric derivatives to support this!

nightdr commented on bingo/local_optimizers/continuous_local_opt.py:
```diff
@@ -179,10 +192,17 @@ def _run_algorithm_for_optimization(self, sub_routine, individual, params):
-                                                args=(individual),
-                                                method=self._algorithm,
-                                                tol=1e-6)
+            if self._algorithm in JACOBIAN_SET:
```
I just added functionality and respective test cases to cover this

nightdr commented:
I added changes to fix up the issues brought up in the initial review.

In terms of other issues that we discussed in this week's meeting:
- I believe the culprit of the "BFGS" and "CG" algorithms with the absolute value case was the gradient tolerance needed for each method (i.e. the infinity norm of the gradients need to fall below a certain threshold, defaulted to 1e-05, for convergence, see https://docs.scipy.org/doc/scipy/reference/optimize.minimize-bfgs.html and https://docs.scipy.org/doc/scipy/reference/optimize.minimize-cg.html for more details). I changed the respective integration and unit test cases to work with functions that have decreasing gradients, but functions with constant gradients might cause issues in the future when using these algorithms.
- In regards to using FitnessFunction properties in the GradientMixin classes, it seems that it is convention to assume that mixins will be paired with some classes and be able to use those classes' properties. Furthermore, I did not change get_jacobian() to return both the fitness vector and the jacobian because it would complicate using get_jacobian() in the scipy root call for continuous local optimization. However, I could add a wrapper function for the scipy root call if you think that it would be beneficial to return both the fitness vector and jacobian in the function.

gbomarito commented on bingo/evaluation/fitness_function.py:
```diff
@@ -98,57 +98,30 @@ def __call__(self, individual):
     @staticmethod
     def _mean_absolute_error_derivative(fitness_vector, fitness_partials):
-        return np.mean(np.sign(fitness_vector) * fitness_partials)
+        return np.mean(np.sign(fitness_vector) * fitness_partials, axis=1)
```
lets move the _metric_derivative functionality to the VectorGradientMixin since I don't remember it being used outside that class

nightdr replied:
sounds good, I implemented this in the most recent changes!

gbomarito reviewed (changes requested):
Looks good, just 1 minor update request before merging

nightdr commented:
I made the changes we discussed during our meeting. However, I left the inheritance order as mixin classes first then the base class. I left the order like this because there is no super call in FitnessFunction I could do to pass a metric to a mixin (i.e. If we inherit the base class then mixin classes, there has to be some super call that passes the parameters from the base class to the mixins. However, this doesn't seem possible without adding a metric parameter to the FitnessFunction constructor). It also seems that in other libraries like Django, mixins are inherited first followed by the base class. However, let me know if you want me to try to figure out how to inherit the base class first then the mixins.
